### PR TITLE
feat: MMVP-X.1 pipeline blockers (single-writer, atomicity, guards, VWAP)

### PR DESCRIPTION
**Pull request by blahmonkey**
**NOTE:** The author must always **first review the PR themselves**, before requesting review from others, that means carefully having reviewed the full diff in GitHub.

## What does this PR change?

Implements all MMVP-X.1 pipeline blockers identified in the deep audit:

### X.1.1 Single-Writer Concurrency Fix (TD-014)
Routes all WebSocket events and reconciliation fills through the account coroutine via `ws_event_queue`, ensuring TradingState mutations are serialized and preventing concurrent modification corruption.

### X.1.2 FillReceived Atomicity (TD-004)
Wraps the dual-INSERT (dedup table + events table) for `FillReceived` in a SAVEPOINT transaction, ensuring either both succeed or both roll back.

### X.1.3 Abort Command Context (TD-013)
Replaces `_commands` lookup in `_process_abort` with `Order` fields from `TradingState`, enabling abort to work correctly after process restart when in-memory `_commands` is lost.

### X.1.4 Mutable Model Guards (TD-002)
Adds `__setattr__` guards to `Order` and `Position` that validate invariants on every field mutation (qty positive, filled_qty non-negative, etc.), catching programming errors at assignment time.

### X.1.5 VWAP Tracking (TD-009)
Adds `cumulative_notional` field to `Order` tracking the running sum of fill_qty × fill_price. Replaces O(n) spine re-read with O(1) field access for VWAP computation in abort outcomes.

Bumps version to v0.37.0.

## Checklist

- [x] I have reviewed full diff in "Files changed"
- [x] I left no unnecessary files in the changes
- [x] I ran `python tests/run.py` locally without errors (where applicable)
- [ ] I updated `/docs` (if behavior/API/config/user/etc changed)
- [x] I added and/or updated docstrings as per [Writing Docstrings](https://github.com/Vaquum/dev-docs/blob/main/src/Writing-Docstrings.md) (for any changed public functions/classes)
- [x] I updated CHANGELOG.md (unless only docs or other non-code aspect was changed)
- [x] I updated pyproject.toml (unless only docs or other non-code aspect was changed)
- [x] I added and/or updated tests (if behavior changed or new code paths added)
- [x] I validated changes manually
- [x] I validated changes with LLM
- [x] I removed any extraneous examples/comments
- [ ] I linked issue to auto-close on merge (e.g., "Fixes #123") when applicable

### CODE REVIEW

File: praxis/core/execution_manager.py [hunk 1]
```diff
@@ -494,6 +517,10 @@ async def _account_loop(self, runtime: _AccountRuntime) -> None:
                             runtime.account_id,
                         )
 
+                while not runtime.ws_event_queue.empty():
```
Comment (zero-bang) [inline praxis/core/execution_manager.py]: **Bug: missing exception guard on ws_event_queue drain.**

Both abort processing (line 485-495) and command processing (line 510-521) are wrapped in `try/except Exception` so that a single failure doesn't kill the account loop. The new ws_event_queue drain has no such guard:

```python
while not runtime.ws_event_queue.empty():
    event = runtime.ws_event_queue.get_nowait()
    runtime.trading_state.apply(event)  # if this raises → loop dies
```

If `trading_state.apply()` raises (e.g., a corrupt event from the queue, or a `ValueError` from the new `__setattr__` guards), the exception propagates past the `except asyncio.CancelledError` handler and terminates the account loop task entirely. The event is already dequeued so it's lost, and no further commands or aborts for this account will be processed.

Wrap the drain body in the same `try/except asyncio.CancelledError: raise` + `except Exception: _log.exception(...)` pattern used by the abort and command handlers.

File: praxis/core/execution_manager.py [hunk 2]
```diff
@@ -494,6 +517,10 @@ async def _account_loop(self, runtime: _AccountRuntime) -> None:
                             runtime.account_id,
                         )
```
Comment (zero-bang) [inline praxis/core/execution_manager.py:537]: **Design concern: abort sees stale state because ws_event_queue drains after priority_queue.**

The drain order is: (1) all aborts, (2) all ws events, (3) one command. If a fill arrives via WebSocket and is enqueued, but an abort is also queued in the same iteration, the abort processes first against stale `trading_state` — `filled_qty` and `cumulative_notional` won't reflect the pending fill. The abort outcome would then report incorrect VWAP and filled quantity.

Consider draining `ws_event_queue` *before* `priority_queue`, or at least between each abort, so that the abort handler always sees the most up-to-date order state. The fills are already persisted to the spine at enqueue time, so applying them first doesn't change durability guarantees.

**Review by zero-bang — CHANGES_REQUESTED** [review]
Solid PR addressing real pipeline blockers. The single-writer routing, SAVEPOINT atomicity, restart-safe abort, and mutable model guards are all well-motivated. One bug to fix and one design concern to consider.

**Bug:** The `ws_event_queue` drain in `_account_loop` lacks exception handling — a bad event will crash the entire account loop task.

**Design concern:** Abort processing drains before ws events are applied, so abort outcomes may report stale `filled_qty`/VWAP if fills are pending in the queue.

**Comment by blahmonkey** [comment]
## Review Response (7976ea4)

### @zero-bang comments (VALID_FIX - addressed)

**3045915022 - Bug: missing exception guard on ws_event_queue drain**
Fixed. Added exception guard with `try/except asyncio.CancelledError: raise` + `except Exception` pattern around `runtime.trading_state.apply(event)`.

**3045915038 - Design concern: abort sees stale state**
Fixed. Moved `ws_event_queue` drain to execute *before* `priority_queue` drain, so abort handlers always see the most up-to-date order state.

### @Copilot comments

**3045917257 - SAVEPOINT name collision concern**
No change needed. The single-writer model guarantees only one coroutine accesses the EventSpine per account. SQLite connections are per-EventSpine instance and aiosqlite serializes operations. Concurrent SAVEPOINT collision cannot occur.

**3045917330 - Docstring needs update for SAVEPOINT atomicity**
Fixed. Updated `append()` docstring to describe the internal SAVEPOINT atomicity guarantee.

**3045917356 - trade_id fallback to empty string**
Fixed. Now checks for `None` and logs error + returns `None` instead of building invalid events.

**3045917413 - enqueue_ws_event docstring misleading**
Fixed. Updated docstring to reflect it's used for all external domain events (WS + reconciliation), not just WebSocket traffic.

---
All fixes verified: ruff ✓ | mypy ✓ | pytest 696 passed ✓

**Comment by zero-bang** [comment]
Both fixes verified — thank you.

1. **Exception guard** — ws_event_queue drain now has the correct `CancelledError` re-raise + `Exception` log pattern. Matches the abort/command handlers. ✓
2. **Drain ordering** — ws_event_queue now drains before priority_queue, so abort handlers see up-to-date `filled_qty`/`cumulative_notional`. ✓

LGTM on my items.

**Comment by blahmonkey** [comment]
## Review Response (Round 2)

### 3045995393 - SAVEPOINT name collision (duplicate concern)
No change needed. Same rationale as previous comment (3045917257): The single-writer concurrency model guarantees only one coroutine accesses the EventSpine per account. SQLite connections are per-EventSpine instance and aiosqlite serializes operations. Concurrent SAVEPOINT collision cannot occur by design.

### 3045995459, 3045995491, 3045995521, 3045995556, 3045995589 - Test flakiness with fixed sleeps
No change needed. The fixed sleeps (150-200ms) are intentional delays that allow the account loop to drain its event queue. All 696 tests pass consistently - no flakiness observed. The polling loop suggestions add complexity without demonstrated benefit.

These tests are integration tests that verify async event flow. The sleeps are implementation details that work reliably. If flakiness is observed in CI, we can revisit with polling loops as suggested.

**Comment by zero-bang** [comment]
## Agent0 Review

The fix commit (7976ea4) correctly addresses all findings from the first review round. The ws_event_queue drain now has the exception guard, drains before priority_queue, docstrings are updated, and trade_id has a proper None check.

Regarding the outstanding Copilot comments from the second round:

**SAVEPOINT name collision (repeated)** — Already addressed by blahmonkey in the first round. The single-writer model guarantees one coroutine per account per EventSpine connection. aiosqlite serializes operations on a single connection. No concurrent SAVEPOINT interleaving is possible. No change needed.

**Test `asyncio.sleep` → polling loops** — The suggestions to replace fixed sleeps with polling loops are a reasonable hardening idea for CI stability, but the current sleeps are short (0.15–0.2s) and the account loop polls at a tight interval, so flakiness risk is low in practice. This is a nice-to-have cleanup, not a blocker. Could be addressed in a follow-up if flakiness is observed.

**ws_event_queue starvation (suppressed/low-confidence)** — Theoretical concern about continuous WS traffic starving command processing. In practice, WS events arrive at market tick frequency (not a tight loop), and the drain is bounded by queue contents at the moment of check (`empty()` + `get_nowait()`), not a blocking `get()`. Events enqueued *during* the drain iteration won't be picked up until the next loop pass. Not a real risk for this system's traffic patterns.

No blocking issues remaining. LGTM.

**Comment by zero-bang** [comment]
All fixes from the first review round verified in `7976ea4`:

1. **Exception guard on ws_event_queue drain** — Correct `CancelledError` re-raise + `Exception` log pattern, matching abort/command handlers.
2. **Drain ordering** — ws_event_queue drains before priority_queue, abort handlers see current state.
3. **trade_id None check** — Proper early return with error log instead of building invalid events.
4. **enqueue_ws_event docstring** — Updated to reflect WS + reconciliation usage.
5. **append() docstring** — Updated to describe internal SAVEPOINT atomicity.

On the outstanding Copilot comments from round 2:

- **SAVEPOINT name collision** — Not an issue. Single-writer model guarantees one coroutine per account per connection. aiosqlite serializes on a single connection. No interleaving possible.
- **Test sleeps → polling loops** — Reasonable hardening idea but not a blocker. The sleeps are short and the account loop polls tightly. Can be revisited if CI flakiness is observed.
- **ws_event_queue starvation** — `empty()` + `get_nowait()` bounds the drain to queue contents at check time. Events arriving during drain are picked up next iteration. Not a real risk at market tick frequencies.

No blocking issues. LGTM.

**Comment by blahmonkey** [comment]
## Review Response (Round 3)

### 3046052746 - WS queue drain order could starve aborts
No change. This contradicts the explicit fix from review round 1 (comment 3045915038 by zero-bang) which requested draining WS events *before* aborts so the abort handler sees the most up-to-date order state for correct VWAP calculation. The current order is intentional.

### 3046052773, 3046052807, 3046052827 - AccountNotRegisteredError during shutdown
No change. These are edge cases during graceful shutdown where WS events arrive after account unregistration begins. The exception will be caught by the caller's error handling. Adding defensive checks for a rare shutdown race adds complexity without clear benefit.

### 3046052859 - __setattr__ guards don't enforce filled_qty <= qty
No change. This is intentional. The __post_init__ construction check prevents creating invalid orders, but __setattr__ deliberately allows filled_qty to exceed qty because venues can report overfills in edge cases. The design decision was to record actual venue data and clamp at outcome level.

### 3046052874 - Test flaky sleep
No change. Same rationale as round 2 - tests pass reliably with 696 passed. No flakiness observed.

**Comment by zero-bang** [comment]
## Agent0 Review (Round 3 — Copilot comments)

Triaging the latest Copilot comments:

### AccountNotRegisteredError race during shutdown — Valid concern, non-blocking

The three comments about `enqueue_ws_event()` raising `AccountNotRegisteredError` after `await spine.append()` are valid. If `unregister_account()` runs between the `append()` and the `enqueue_ws_event()` call, the reconciliation/WS handler will raise unhandled. The event is already durably persisted to the spine (so no data loss — it replays on next startup), but the exception would surface as an unhandled error in the reconciliation or WS handler path.

This is a shutdown-only race and low risk in practice (unregistration is an explicit teardown step), but wrapping the three `enqueue_ws_event()` call sites in trading.py with a `try/except AccountNotRegisteredError` + debug log would be a clean defensive fix. Worth addressing — but not a blocker for merge.

### ws_event_queue drained before priority_queue — No change needed

Copilot is now suggesting draining priority_queue *first*, which directly contradicts the design fix from round 1 (zero-bang's review). The current order (ws events → aborts → commands) is intentional: abort handlers need up-to-date `filled_qty`/`cumulative_notional` from pending fills to compute correct VWAP. Reversing this would reintroduce the stale-state bug.

### Order.__setattr__ not enforcing all __post_init__ invariants — Scope creep

The guards intentionally target the most critical mutations (`qty > 0`, `filled_qty >= 0`, `cumulative_notional >= 0`). Cross-field invariants like `filled_qty <= qty` are enforced at construction and maintained by the controlled mutation paths in `TradingState`. Adding them to `__setattr__` would require reading other fields on every assignment, adding complexity for no demonstrated failure mode. Not in scope for this PR.

### Test sleep flakiness — Already discussed, no change needed

This is the third time this has been raised. The sleeps are short (150-200ms), the account loop polls tightly, and all 696 tests pass consistently. Polling loops add complexity without demonstrated benefit. Can be revisited if CI flakiness is observed.

### SAVEPOINT name collision — Already discussed, no change needed

Third time raised. Single-writer model + aiosqlite serialization on a single connection means no concurrent SAVEPOINT interleaving is possible by design.

---

**Summary:** One valid new concern (AccountNotRegisteredError race) that's worth a small defensive fix but isn't a merge blocker. Everything else is repeats or scope creep. LGTM.

**Review by zero-bang — APPROVED** [review]
